earthfile2llb/listener.go

```diff
@@ -1091,3 +1100,69 @@ func unescapeSlashPlus(str string) string {
 	// TODO: This is not entirely correct in a string like "\\\\+".
 	return strings.ReplaceAll(str, "\\+", "+")
 }
+
+type matrixEntry struct {
+	key    string
+	values []*string
+}
+
+func buildArgMatrix(args []string) ([][]string, error) {
+	matrix := make([]matrixEntry, 0, len(args))
+	for _, arg := range args {
+		k, v, err := parseKeyValue(arg)
+		if err != nil {
+			return nil, err
+		}
+
+		found := false
+		for i, e := range matrix {
```

Comment:
nit: maybe use a map here instead so it's nlogn rather than n^2; but n is so small it doesn't really matter.

Reply:
another small nit, maybe call `matrix` `groupedArgs` instead? as it's not really a matrix.

Reply:
Corey had a similar remark about using maps. I've not used maps due to the ordering not being guaranteed in a map, which can cause cache-busting.

Good call on the naming.